service/src/test/java/bio/terra/workspace/service/resource/controlled/ControlledResourceServiceNotebookTest.java

```diff
@@ -0,0 +1,768 @@
+    petSaService.disablePetServiceAccountImpersonation(
+        workspaceId, user.getEmail(), user.getAuthenticatedRequest());
+    IamCow userIamCow = crlService.getIamCow(user.getAuthenticatedRequest());
+    // Assert the user does not have access to their pet SA before the flight
+    // Note this uses user credentials for the IAM cow to validate the user's access.
+    GcpCloudUtils.runWithRetryOnException(
+        () ->
+            throwIfImpersonateSa(
+                ServiceAccountName.builder()
+                    .projectId(projectId)
+                    .email(serviceAccountEmail)
+                    .build(),
+                userIamCow));
+
+    String jobId =
+        controlledResourceService.createAiNotebookInstance(
+            resource,
+            creationParameters,
+            DEFAULT_ROLE,
+            new ApiJobControl().id(UUID.randomUUID().toString()),
+            "fakeResultPath",
+            user.getAuthenticatedRequest());
+    jobService.waitForJob(jobId);
+    assertEquals(
+        FlightStatus.ERROR, stairwayComponent.get().getFlightState(jobId).getFlightStatus());
+
+    assertNotFound(resource.toInstanceName(projectId), crlService.getAIPlatformNotebooksCow());
+    assertThrows(
+        ResourceNotFoundException.class,
+        () ->
+            controlledResourceService.getControlledResource(
+                resource.getWorkspaceId(), resource.getResourceId()));
+    // This check relies on cloud IAM propagation and is sometimes delayed.
+    GcpCloudUtils.runWithRetryOnException(
+        () ->
+            throwIfImpersonateSa(
+                ServiceAccountName.builder()
+                    .projectId(projectId)
+                    .email(serviceAccountEmail)
+                    .build(),
+                userIamCow));
+    // Run and undo the flight again, this time triggered by the default user's pet SA instead of
+    // the default user themselves. This should behave the same as the flight triggered by the
+    // end-user credentials but has historically hidden bugs, so is worth testing explicitly.
+    AuthenticatedUserRequest petCredentials =
+        petSaService.getWorkspacePetCredentials(workspaceId, user.getAuthenticatedRequest()).get();
+    String petJobId =
+        controlledResourceService.createAiNotebookInstance(
+            resource,
+            creationParameters,
+            DEFAULT_ROLE,
+            new ApiJobControl().id(UUID.randomUUID().toString()),
+            "fakeResultPath",
+            petCredentials);
+    jobService.waitForJob(petJobId);
+    // Confirm this flight status is ERROR, not FATAL.
+    assertEquals(
+        FlightStatus.ERROR, stairwayComponent.get().getFlightState(petJobId).getFlightStatus());
+  }
+
+  @Test
+  @DisabledIfEnvironmentVariable(named = "TEST_ENV", matches = BUFFER_SERVICE_DISABLED_ENVS_REG_EX)
+  void updateAiNotebookResourceDo() throws InterruptedException, IOException {
```

Comment:
out of scope for this PR, but I wonder if we could shrink test time more by combining some tests to use the same instance - e.g. create the notebook and test do steps, test do/undo updates on that instance, etc.

Reply:
Yes. In general, we should do a pass over our tests to find those cases and ones where multiple tests are exercising the same code paths.